pkg/securitypolicy/securitypolicyenforcer_rego.go

```diff
@@ -21,50 +36,491 @@ func init() {
+			return err
+		}
+
+		var end string
+		if i < len(containers)-1 {
+			end = ",\n"
+		} else {
+			end = "\n"
+		}
+
+		if _, err := builder.WriteString(end); err != nil {
+			return err
+		}
+	}
+
+	if _, err := builder.WriteString("]\n"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p securityPolicyInternal) MarshalRego() (string, error) {
+	builder := new(strings.Builder)
+	if _, err := builder.WriteString(fmt.Sprintf("package policy\nallow_all := %v\n", p.AllowAll)); err != nil {
+		return "", err
+	}
+
+	if err := addContainers(builder, p.Containers); err != nil {
+		return "", err
+	}
+
+	return builder.String(), nil
+}
+
+func createRegoEnforcer(state SecurityPolicyState, defaultMounts []oci.Mount, privilegedMounts []oci.Mount) (SecurityPolicyEnforcer, error) {
+	policy, err := state.SecurityPolicy.toInternal()
+	if err != nil {
+		return nil, fmt.Errorf("error converting to internal format: %w", err)
+	}
+
+	regoPolicy, err := newRegoPolicyFromInternal(policy, defaultMounts, privilegedMounts)
+	if err != nil {
+		return nil, fmt.Errorf("error converting to Rego: %w", err)
+	}
+
+	regoPolicy.base64policy = state.EncodedSecurityPolicy.SecurityPolicy
+	return regoPolicy, nil
+}
+
+func newRegoPolicyFromInternal(securityPolicy *securityPolicyInternal, defaultMounts []oci.Mount, privilegedMounts []oci.Mount) (*RegoEnforcer, error) {
+	policy := new(RegoEnforcer)
+	policy.behavior = PolicyCode
+	if code, err := securityPolicy.MarshalRego(); err == nil {
+		policy.objects = code
+	} else {
+		return nil, fmt.Errorf("failed to convert json to rego: %w", err)
+	}
+
+	defaultMountData := make([]interface{}, 0, len(defaultMounts))
+	privilegedMountData := make([]interface{}, 0, len(privilegedMounts))
+	policy.data = map[string]interface{}{
+		"started":          []string{},
+		"devices":          map[string]string{},
+		"containers":       map[string]interface{}{},
+		"defaultMounts":    appendMountData(defaultMountData, defaultMounts),
+		"privilegedMounts": appendMountData(privilegedMountData, privilegedMounts),
+		"sandboxPrefix":    guestpath.SandboxMountPrefix,
+		"hugePagesPrefix":  guestpath.HugePagesMountPrefix,
+	}
+	policy.mutex = new(sync.Mutex)
+	policy.base64policy = ""
+
+	modules := map[string]string{
+		"behavior.rego":  policy.behavior,
+		"objects.rego":   policy.objects,
+		"framework.rego": FrameworkCode,
+	}
+
+	// TODO temporary hack for debugging policies until GCS logging design
+	// and implementation is finalized. This option should be changed to
+	// "true" if debugging is desired.
+	options := ast.CompileOpts{
+		EnablePrintStatements: false,
+	}
+
+	if compiled, err := ast.CompileModulesWithOpt(modules, options); err == nil {
+		policy.compiledModules = compiled
+	} else {
+		return nil, fmt.Errorf("rego compilation failed: %w", err)
+	}
+
+	return policy, nil
+}
+
+func (policy RegoEnforcer) Query(input map[string]interface{}) (rego.ResultSet, error) {
+	store := inmem.NewFromObject(policy.data)
```

Comment:
Is it possible for a policy evaluation to persist data into the store, and thus write new values into `policy.data`? Do we need to worry about data being persisted over time, leading to increased memory usage?

Reply:
We add data to policy.data over the course of a run to track information that is needed to make policy decisions. I'm not sure what you mean by persisting data into the store and thus write new values into policy.data.

Data is persisted (in policy.data), there will be some memory growth, same as with the standard enforcer.

Reply:
Taking a fresh look at this, I guess `store` is re-generated from `policy.data` for every `Query` call, used for the query evaluation, and then discarded. Would it be better to instead persist `store` on `policy` so that it is re-used every time?

Reply:
Store is (from my reading of the OPA Go code) intended to be a readonly object, and should be recreated every time the data changes. As the data changes quite a bit, I made the choice to produce it every time a query happens. That said, we could have it as member of `policy` that gets set to nil every time the data changes. I'm not sure how much this saves, however.

Reply:
I think your reasoning for the current way makes sense. Thanks for the context.